File: Chapter01/netconf.js
#!/usr/bin/env node

const util = require("util");
const child_process = require('child_process');
const xml2js = require('xml2js');

const DELIMITER="]]>]]>";

const xmlRpcCommand = function(command) {
        return [
                "<rpc>\n",
                "<command format=\"text\">\n",
                command,
                "</command>",
                "</rpc>\n",
                DELIMITER,
                "\n"
        ];
};

var walk = function(obj, path) {
	var result = obj;
	path.forEach(function (cur, ind, array) {
		if (result) result=result[cur];
	});
	return result;
}


if (process.argv.length!=4) {
	console.warn("Usage: netconf.js user@hostname command\n");
	process.exit(1);
}
var hostname = process.argv[2];
var command = process.argv[3];

var child = child_process.spawn(
	"/usr/bin/ssh", [
        "auto@"+hostname,
        "-q",
		"-p", "830",
		"-i", "junos_auto_id_rsa",
		"-s", "netconf"
	]
);

var data="";


child.stderr.on('data', function(chunk) { process.stderr.write(chunk, "utf8"); });
child.stdout.on('data', function(chunk) {
	data+=chunk;
	if ((index=data.indexOf(DELIMITER))!=-1) {
		var xml = data.slice(0, index);
		data = data.slice(index + DELIMITER.length);
		xml2js.parseString(xml, function(err, result) {
			if (err) throw err;
			if (result['hello']) return;
			if (output=walk(result, ['rpc-reply', 'output', 0])) {
				console.log(output);
			} else if (config=walk(result, ['rpc-reply', 'configuration-information', 0, 'configuration-output', 0])) {
				console.log(config);
			} else if (error=walk(result, ['rpc-reply', 'rpc-error', 0, 'error-message', 0])) {
				console.log(error);
			} else {
				console.log("Unexpected empty response");
			}
			child.stdin.end();
		});
	}
});

child.on('error', function(err) { console.log("SSH client error: ", err); })

xmlRpcCommand(command).forEach(function(cur, ind, array) { child.stdin.write(cur, "utf8")} );

File: Chapter01/netconf.py
#!/usr/bin/env python

import sys
import subprocess
import xml.etree.ElementTree as ET

class NETCONFClient(object):

	DELIMITER = ']]>]]>\n'

	def __init__(self, hostname):
		self.ssh = subprocess.Popen([
				"/usr/bin/ssh",
				"-q",
				"-i", "junos_auto_id_rsa",
				"-p", "830",
				"-s", 
				hostname,
				"netconf",
			],
			stdin=subprocess.PIPE,
			stdout=subprocess.PIPE)

	def __del__(self):
		self.ssh.stdin.close()

	def read(self):
		data=""
		for line in iter(self.ssh.stdout.readline, NETCONFClient.DELIMITER):
			if line=='':
				raise IOError("ssh session ended unexpectedly")
			data += line

		return ET.fromstring(data)

	def cmdrpc(self, cmd):
		e = ET.Element("rpc")
		e.append(ET.Element("command", {'format': "text"}))
		e.find("command").text = cmd;
		self.ssh.stdin.write(ET.tostring(e))
		self.ssh.stdin.write(NETCONFClient.DELIMITER)

# Error handling
def onError(exception_type, exception, traceback):
    print "%s: %s" % (exception_type.__name__, exception)

sys.excepthook = onError

# Read command line
if len(sys.argv) < 3:
	print "Usage: netconf.py hostname command"
	sys.exit(1)

netconf = NETCONFClient("auto@"+str(sys.argv[1]))
response = netconf.read()
netconf.cmdrpc(" ".join(sys.argv[2:]))
response = netconf.read()

output = response.find(".//{urn:ietf:params:xml:ns:netconf:base:1.0}output")
config = response.find(".//{urn:ietf:params:xml:ns:netconf:base:1.0}configuration-output")
error = response.find(".//{urn:ietf:params:xml:ns:netconf:base:1.0}error-message")

if output != None:
	print output.text
elif config != None:
	print config.text
elif error != None:
	print error.text
else:
	print "NETCONF server provided no usable response"
